src/common/libflux/reactor.c

```diff
@@ -113,6 +116,42 @@ flux_reactor_t *flux_get_reactor (flux_t *h)
     return r;
 }
 
+static __thread double timeout_seconds = 0;
+
+static void check_timeouts(struct ev_loop *current_loop, int t, void *vw) {
+  ev_watcher *w = vw;
+  ev_tstamp remaining = 0;
+  if (t == EV_TIMER)
+    remaining = ev_timer_remaining(current_loop, (ev_timer *)w);
+  else if (t == EV_PERIODIC)
+    remaining = ev_periodic_at(w) - ev_now(current_loop);
+
+  // take the minimum
+  if (isnan(timeout_seconds) || timeout_seconds > remaining)
+    timeout_seconds = remaining;
+}
+
+static __thread bool any_active = false;
+
+static void check_any_active(struct ev_loop *current_loop, int t, void *vw) {
+  ev_watcher *w = vw;
+  if (ev_is_active(w))
+    any_active = true;
+}
+
+double flux_reactor_resume_timeout(flux_reactor_t *r) {
+  any_active = false;
+  ev_walk(r->loop, EV_CHECK | EV_IDLE | EV_PREPARE, check_any_active);
```

Comment:
Really excellent job finding the undocumented `ev_walk()` interface @trws!  Apparently you looked at libev code and did not go blind!

I think I may have misspoken in our chat - I believe only the `ev_idle` watcher prevents the loop from blocking when active.  The prepare and check watchers can remain active and the loop will still block.  So I think you'd want to drop EV_CHECK and EV_PREPAR from the `ev_walk()` call highlighted above.  (I believe this makes it more correct although in the particular case of the check watcher in the future continuation implementation, it looks like we don't leave it running when we expect the loop to block, so I don't expect it will make a difference in testing)

If I understand correctly, the strategy is to never block in `ev_run()`, so the only event sources that will cause the flux loop to run are the timeout and the `flux_pollfd()`.  So that really means _none_ of the other watchers besides the flux_t handle and the idle watcher will work as asynchronous event sources in the flux reactor.   For example a file descriptor watcher in the flux reactor cannot make the parent event loop wake up.  Have I got that right?  It seems like your description is saying otherwise.

Due to the limitations, regardless of what they are, I was suggesting in the chat that we maybe build a flag in to the flux reactor to indicate to `flux_reactor_create()` that it's running the special "adjunct reactor" mode.  If the flag is set, fail any attempt to create watchers that are unsupported.  Conversely, fail `flux_reactor_resume_timeout()` if the flag is clear.  Your proposed interface appears to be complete for embedding a `flux_t` in a foreign loop.  However if someone decided to use some lesser used API functions, libsubprocess, for example that makes use of child watchers and buffer watchers (which internally use fd watchers), it would just fail and there would be no intermittent hangs to track down.

When we get a little further along, it would probably make sense to add a unit test to libflux/test that embeds the flux reactor in...another flux reactor (or in libev)?

This is real progress, thanks so much for pushing it forward!!

Reply:
Thanks @garlick, and you're absolutely right about the support, for some reason I was thinking about using the libev backend_fd as the watched file descriptor rather than a pollfd and messed up what will and will not work.  Using this interface along with a pollfd will get you the messages and associated continuations, timers, prepare, check and idle watchers on the underlying reactor.  None of the others will work without some additional work.  If we're willing to extend libev a bit to expose the backend_fd, everything else would work when using epoll on linux, with some unfortunate properties for signalfd.  I'm not sure if I want to go that far, at least for this step, so we'll be limited to a handle, messages, continuations and timers.

With that being the limitation, I'm considering how we should build the interface to mark a reactor in "embedded" mode.  It could be directly a property of trying to use `flux_reactor_resume_timeout`, checking for any unsupported watcher and bailing out or we could make it happen when a pollfd is requested from the handle but I'm not sure if we could actually do it on create of the reactor.  Mainly I'm thinking of the situation where a handle gets created in bindings, and we don't know if the user intends to use it directly with `reactor_run` in blocking mode or if they intend to embed it.  What do you think?

Reply:
Performing the `ev_walk()` sweep each time  `flux_reactor_resume_timeout()` is called and skipping the flag entirely seems appealing since it ought to catch any rogue watchers added to the event loop, and makes this solution self contained.  Although on a critical path of sorts, that scan ought to be quick since in this use case there's likely to only be a handful of watchers.

One thing to note (not necessarily a problem but might come up in test) is that the sweep won't catch watchers that are created but not started.  Libev doesn't track watchers that aren't started.

Upon looking at reactor code i was regretting suggesting we test a flag in the watcher create path.  it seems like we'd either have to test  it in every watcher implementation or refactor the generic watcher interface.  And we've exposed that interface in the public API.

Reply:
Yeah, I think making it self contained is worth the possible cost at least for now.  If we later found it was prohibitive for some reason we could re-work without breaking anything, either by tracking a different way or adding something to libev to have a more efficient walk that doesn't require thread-local memory, but I don't expect it to be performance critical or terribly expensive.

I was slightly surprised by the behavior of non-started watchers being missing at first, but in the end I think it actually helps us to some extent.  This way if something creates a watcher because it might use it later, but doesn't actually use it, we aren't pessimizing and breaking on that case. It also means we should be iterating over less watchers.

Reply:
Since there's already an `ev_walk()` in there to look for idle watchers, maybe the callback could just be extended to also look for the rogue watchers.

Using the callback`s `void *` argument to point to a struct on the stack of `flux_reactor_resume_timeout()` could let you pass in the watcher and flags and avoid thread-local storage for flags, unless I'm missing something.

Reply:
ev_walk doesn't offer a void * argument in its callback that we can control unfortunately, I originally thought it did but it passes in the watcher there and expects the callback to determine the type from the int.  If we had everything we need in the watcher's `data` pointer that would work, but it puts us in a chicken and egg situation.   Actually... I could save the data pointer from the loop, put a stack variable in there for state that will get passed in, then restore it.  That might actually work, will give it a shot.

Reply:
Ah I misread that, sorry.

`data` is member of `flux_watcher_t` not `ev_watcher`, unless I'm not spotting another `data` member, as it's not all defined in one place it seems.

Probably no big deal if we have to stick with the thread local storage, at least for nw.